Source/AMRInterpolator/SurfaceExtraction.hpp

```diff
@@ -0,0 +1,201 @@
+    {
+        int num_surfaces; //!< number of surfaces over which to extraction
+        std::vector<double>
+            surface_param_values; //!< the values of the
+                                  //!< parameter that gives the required
+                                  //!< surfaces with SurfaceGeom geometry (e.g.
+                                  //!< radii for spherical shells)
+        int num_points_u; //!< the number of points for the first parameter
+                          //!< that parameterises each surface
+        int num_points_v; //!< the number of points for the second parameter
+                          //!< that parameterises each surfaces
+        std::vector<int> extraction_levels; //!< the level on which to do the
+                                            //!< extraction for each surface
+        bool write_extraction; //!< whether or not to write the extracted data
+
+        int min_extraction_level()
+        {
+            return *(std::min_element(extraction_levels.begin(),
+                                      extraction_levels.end()));
+        }
+    };
+
+  protected:
+    const SurfaceGeometry *m_geom_ptr; //!< the pointer to the geometry class
+    const params_t m_params;
+    std::vector<std::pair<int, Derivative>> m_vars; //!< the vector of pairs of
+    //!< variables and derivatives to extract
+    const double m_dt;
+    const double m_time;
+    const bool m_first_step;
+    const double m_restart_time;
+    const int m_num_points; //!< the total number of points per surface
+    const double m_du;      //!< the grid spacing in u (used in integrate)
+    const double m_dv;      //!< the grid spacing in v (used in integrate)
+
+    std::vector<std::vector<double>> m_interp_data;
+    std::array<std::vector<double>, CH_SPACEDIM> m_interp_coords;
+
+    bool m_done_extraction; //!< whether or not the extract function has been
+                            //!< called for this object
+
+    //! returns the flattened index for m_interp_data and m_interp_coords
+    //! associated to given surface, u and v indices
+    int index(int a_isurface, int a_iu, int a_iv) const
+    {
+        return a_isurface * m_num_points + a_iu * m_params.num_points_v + a_iv;
+    }
+
+  public:
+    //! Normal constructor which requires vars to be added after construction
+    //! using add_var or add_vars
+    SurfaceExtraction(SurfaceGeometry *a_geom_ptr, params_t &a_params,
+                      double a_dt, double a_time, bool a_first_step,
+                      double a_restart_time = 0.0);
+
+    //! add a single variable or derivative of variable
+    void add_var(const int a_var, const Derivative a_deriv = Derivative::LOCAL);
+
+    //! add a vector of variables/derivatives of variables
+    void add_vars(const std::vector<std::pair<int, Derivative>> &a_vars);
+
+    //! Alternative constructor with a predefined vector of variables
+    SurfaceExtraction(SurfaceGeometry *a_geom_ptr, params_t &a_params,
```

Comment:
Feature request: can we have a 3rd constructor in SurfaceExtraction that receives not a 'const std::vector<std::pair<int, Derivative>> &a_vars', but a 'const std::vector<int> &a_vars' that assumes no derivatives? So that I could make SurfaceExtraction(..., params, {c_chi, c_phi}, ...) without worrying about derivatives if I want none

Reply:
Done, also added a corresponding `SurfaceExtraction::add_vars(const std::vector<int> &a_vars)` overload which is used by the new constructor.